### PR TITLE
Adds: dry-run, sync subset, and Fig autocomplete

### PR DESCRIPTION
**Pull request by jazanne**
### Adds new options

- `--dry-run` - preview changes
- `--verbose` - log output setting
- `-f`, `--flag` - sync one flag
- `-T`,  `--tag` - sync flags with tag(s)

### Adds support for [Fig autocomplete](https://fig.io/)

Note, some options like `isRepeatable` work for `--projects`, but not `-p`. I kept them since it is presumably a bug that will be fixed.

### Support ldc config

- use a config file (same format as LaunchDarkly-Labs/ldc) to easily autocomplete api tokens and app uri

### Fixes

- Errors not thrown correctly for non-existent environments

## Testing

Download [Fig](https://fig.io/)

Navigate to this repo/branch in terminal

Run script with `./sync-ld-flags` to get Fig autocompletions

_See README for more details on using script_

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
...
+   ```
+
    **Note:** The environment key can be found in Account Settings within the LaunchDarkly application immediately under the name of the desired environment.
    
    Optionally pass a host override:
 
-        node index.js -p PROJECT_KEY -s SOURCE_ENVIRONMENT_KEY -d DEST_ENVIRONMENT_KEY -t API_TOKEN -H https://your-launch-darkly-deploy.com
+   ```
+   ./sync-ld-flags --api-token API_TOKEN \
+     -H https://your-launch-darkly-deploy.com
```
Comment (mmrj) [inline README.md:45]: Will the user need single quotes around the URL here?

File: README.md [hunk 2]
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
...
+
+```
+./sync-ld-flags [options]
+```
+
+Options:
+
+```
+  -p, --project-key <key>      Project key
+  -s, --source-env <key>       Source environment
```
Comment (mmrj) [inline README.md:65]:
```suggestion
  -s, --source-env <key>       Source environment key
```

File: README.md (hunk 1 above)
Comment (jazanne) [reply README.md:45]: nope, should be ok without it

File: README.md [hunk 3]
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
...
+
+Add `--api-token` and (optionally) `--host` at the beginning of your commands and Fig will provide suggestions for your project and environment keys.
+
+Want to set these configuration variables more easily?
+
+![Auto-complete configuration](docs/images/autocomplete-config.gif)
+
+Auto-complete keys and servers by using a config file:
+
+Add the following configuration to `~/.config/ldc.json`
```
Comment (mmrj) [inline README.md:93]:
```suggestion
You can autocomplete keys and servers by using a config file. Add the following configuration to `~/.config/ldc.json`:
```

File: README.md [hunk 4]
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
...
+
+## Other
+
+This project supports [Fig](https://fig.io/) autocomplete!
+
+Add `--api-token` and (optionally) `--host` at the beginning of your commands and Fig will provide suggestions for your project and environment keys.
+
+Want to set these configuration variables more easily?
+
+![Auto-complete configuration](docs/images/autocomplete-config.gif)
```
Comment (mmrj) [inline README.md:89]:
```suggestion
![Autocomplete configuration.](docs/images/autocomplete-config.gif)
```

File: example-config.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+    "<REPLACE WITH NAME>": {
+        "apitoken": "<REPLACE WITH PERSONAL ACCESS TOKEN>",
+        "server": "<REMOVE OR REPLACE WITH APP URI>"
+    }
+}
```
Comment (jazanne) [inline example-config.json:6]:
```suggestion
}

```

File: index.js [hunk 6]
```diff
@@ -7,6 +7,9 @@ const jsonpatch = require('fast-json-patch');
 const request = require('request');
```
Comment (yusinto) [inline index.js:7]: I think casing consistency matters so in this case should `jsonpatch` be `jsonPatch` instead?

File: index.js [hunk 7]
```diff
@@ -7,6 +7,9 @@ const jsonpatch = require('fast-json-patch');
...
 const program = require('commander');
 
+// Use to calculcate changing flags
```
Comment (yusinto) [inline index.js:10]: Typo `calculcate` => `calculate`

File: index.js [hunk 8]
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
...
-  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken } = config;
+  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken, tags, flag } = config;
+  let isSingle = flag ? true : false;
+  let url = isSingle
+    ? `${baseUrl}/flags/${projectKey}/${flag}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}`
+    : `${baseUrl}/flags/${projectKey}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}${tags ? ("&filter=tags:" + tags.join('+')) : '' }`;
+
   const requestOptions = {
-    url: `${baseUrl}/flags/${projectKey}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}`,
+    url: url,
```
Comment (yusinto) [inline index.js:41]: This can be simplified to:
```ts
const requestOptions = {
    url,
```

File: index.js [hunk 9]
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
 }
 
 const fetchFlags = function (config, cb) {
-  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken } = config;
+  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken, tags, flag } = config;
+  let isSingle = flag ? true : false;
```
Comment (yusinto) [inline index.js:35]: This can be simplified to:

```ts
let isSingle = !!flag
```

File: index.js [hunk 10]
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
...
+  let url = isSingle
+    ? `${baseUrl}/flags/${projectKey}/${flag}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}`
+    : `${baseUrl}/flags/${projectKey}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}${tags ? ("&filter=tags:" + tags.join('+')) : '' }`;
```
Comment (yusinto) [inline index.js:38]: Nested ternaries are quite difficult to read and debug so much so [eslint has a rule disallowing it](https://eslint.org/docs/rules/no-nested-ternary). I recommend we avoid nesting ternaries like this and be more verbose using if/else for example.

File: .fig/autocomplete/src/sync-ld-flags.ts [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      description: "Sync flags with the specified tag(s). Only flags with all tags will sync.",
+      icon: ICON_TAG,
+      exclusiveOn: ["-f", "--flag"],
+      args: {
+        name: "string",
+        description: "Tag name",
+      },
+    },
+    {
+      name: ["--dry-run"],
```
Comment (yusinto) [inline .fig/autocomplete/src/sync-ld-flags.ts:266]: Any reason why there's no shortcode for `dry-run`?

**Review by yusinto — CHANGES_REQUESTED** [review]

File: index.js (hunk 10 above)
Comment (jazanne) [reply index.js:38]: good call - this was hard to follow

File: .fig/autocomplete/src/sync-ld-flags.ts (hunk 11 above)
Comment (jazanne) [reply .fig/autocomplete/src/sync-ld-flags.ts:266]: i wasn't sure what a good shortcode would be since we already use `-d` and `-D` for other opts and i couldn't find any examples of what people typically use for this - open to suggestions!

File: index.js (hunk 6 above)
Comment (jazanne) [reply index.js:7]: hum i don't think it makes a difference since we don't use the latter anywhere that i'm aware of and the module itself uses all lowercase in examples https://www.npmjs.com/package/fast-json-patch

File: index.js (hunk 9 above)
Comment (jazanne) [reply index.js]: i modified this a but to make sure tags aren't included (which they shouldn't be)

File: .fig/autocomplete/src/sync-ld-flags.ts (hunk 11 above)
Comment (yusinto) [reply .fig/autocomplete/src/sync-ld-flags.ts:266]: I looked around, both `make` and `yarn` uses `-n` as an alias for `--dry-run` so we can follow that.

**Comment by yusinto** [comment]
I would like to report a crash which I discovered, when I ran the command using the `-f` param with a non-existent flag:

```bash
./sync-ld-flags --dry-run --api-token api-xxx --project-key string --source-env production --destination-env test -v -f nonExistentFlagKey

/Users/yusngadiman/projects/sync-ld-flags-j/index.js:169
      if (matches.groups && matches.groups.envKey) {
                  ^

TypeError: Cannot read properties of null (reading 'groups')
    at /Users/yusngadiman/projects/sync-ld-flags-j/index.js:169:19
    at Request.callback [as _callback] (/Users/yusngadiman/projects/sync-ld-flags-j/index.js:62:7)
    at Request.self.callback (/Users/yusngadiman/projects/sync-ld-flags-j/node_modules/request/request.js:185:22)

```

**Review by yusinto — CHANGES_REQUESTED** [review]

File: .fig/autocomplete/package.json [hunk 12]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "boilerplate-fig-autocomplete",
+  "version": "1.0.0",
+  "description": "Boilerplate Fig autocomplete spec package",
```
[the remark this answers is not in the document]
Comment (jazanne) [reply .fig/autocomplete/package.json:4]: My thought is no, because this is actually a package.json for the [Fig dev tools](https://fig.io/docs/guides/private-autocomplete), not our own package.

open to changing it though if you think we should

File: index.js (hunk 6 above)
Comment (jazanne) [reply index.js:7]: AH! ok that makes sense - i can definitely change that

**Comment by jazanne** [comment]
@yusinto I fixed the unhandled flag error (404) you found

I also added some basic linting since i realized we didn't have it when i made the `jsonPatch` change - lmk what you think. I can also revert and handle in separate PR if you think that's better

**Comment by yusinto** [comment]
@jazanne linting is a great idea. Thanks for doing this!

**Review by yusinto — APPROVED** [review]
